File: zy1156/backend/src/models/Task.js
const { v4: uuidv4 } = require('uuid');
const { runQuery } = require('../config/database');

class Task {
  constructor({ id, name, description, status, created_at, updated_at }) {
    this.id = id || uuidv4();
    this.name = name;
    this.description = description || '';
    this.status = status || 'pending';
    this.createdAt = created_at;
    this.updatedAt = updated_at;
  }

  static create({ name, description }) {
    const task = new Task({ name, description });
    
    runQuery(
      'INSERT INTO tasks (id, name, description, status, created_at, updated_at) VALUES (?, ?, ?, ?, CURRENT_TIMESTAMP, CURRENT_TIMESTAMP)',
      [task.id, task.name, task.description, task.status]
    );
    
    return task;
  }

  static findById(id) {
    const results = runQuery('SELECT * FROM tasks WHERE id = ?', [id]);
    if (results.length === 0) return null;
    return new Task(results[0]);
  }

  static findAll(limit = 50, offset = 0) {
    const results = runQuery(
      'SELECT * FROM tasks ORDER BY created_at DESC LIMIT ? OFFSET ?',
      [limit, offset]
    );
    return results.map(row => new Task(row));
  }

  static count() {
    const result = runQuery('SELECT COUNT(*) as count FROM tasks');
    return result[0]?.count || 0;
  }

  update({ name, description, status }) {
    const updates = [];
    const values = [];
    
    if (name !== undefined) {
      updates.push('name = ?');
      values.push(name);
      this.name = name;
    }
    if (description !== undefined) {
      updates.push('description = ?');
      values.push(description);
      this.description = description;
    }
    if (status !== undefined) {
      updates.push('status = ?');
      values.push(status);
      this.status = status;
    }
    
    if (updates.length === 0) return this;
    
    updates.push('updated_at = CURRENT_TIMESTAMP');
    values.push(this.id);
    
    runQuery(
      `UPDATE tasks SET ${updates.join(', ')} WHERE id = ?`,
      values
    );
    
    return this;
  }

  delete() {
    runQuery('DELETE FROM tasks WHERE id = ?', [this.id]);
  }

  toJSON() {
    return {
      id: this.id,
      name: this.name,
      description: this.description,
      status: this.status,
      createdAt: this.createdAt,
      updatedAt: this.updatedAt
    };
  }
}

module.exports = Task;

File: zy10382/internal/repository/task.go
package repository

import (
	"customer-probe-api/internal/models"
	"customer-probe-api/pkg/database"
	"customer-probe-api/pkg/utils"
	"time"

	"gorm.io/gorm"
)

type TaskRepository struct{}

func NewTaskRepository() *TaskRepository {
	return &TaskRepository{}
}

func (r *TaskRepository) CreateWithIdempotency(task *models.ProbeTask, idempotencyKey string) (*models.ProbeTask, bool, error) {
	var existingTask models.ProbeTask
	err := database.GetDB().Where("idempotency_key = ?", idempotencyKey).First(&existingTask).Error
	if err == nil {
		return &existingTask, true, nil
	}

	task.IdempotencyKey = idempotencyKey
	if task.ID == "" {
		task.ID = utils.GenerateID()
	}
	if task.Status == "" {
		task.Status = models.TaskStatusPending
	}

	if err := database.GetDB().Create(task).Error; err != nil {
		return nil, false, err
	}
	return task, false, nil
}

func (r *TaskRepository) GetByID(id string) (*models.ProbeTask, error) {
	var task models.ProbeTask
	if err := database.GetDB().First(&task, "id = ?", id).Error; err != nil {
		return nil, err
	}
	return &task, nil
}

func (r *TaskRepository) GetByIdempotencyKey(key string) (*models.ProbeTask, error) {
	var task models.ProbeTask
	if err := database.GetDB().Where("idempotency_key = ?", key).First(&task).Error; err != nil {
		return nil, err
	}
	return &task, nil
}

func (r *TaskRepository) UpdateStatus(taskID, status, errorMsg string) error {
	updates := map[string]interface{}{
		"status":     status,
		"updated_at": time.Now(),
	}
	if errorMsg != "" {
		updates["error_msg"] = errorMsg
	}
	if status == models.TaskStatusRunning {
		updates["started_at"] = time.Now()
	}
	if status == models.TaskStatusCompleted || status == models.TaskStatusFailed || status == models.TaskStatusTimeout {
		updates["completed_at"] = time.Now()
	}
	return database.GetDB().Model(&models.ProbeTask{}).Where("id = ?", taskID).Updates(updates).Error
}

func (r *TaskRepository) AssignTask(taskID, agent string) error {
	return database.GetDB().Model(&models.ProbeTask{}).
		Where("id = ? AND status = ?", taskID, models.TaskStatusPending).
		Updates(map[string]interface{}{
			"status":        models.TaskStatusAssigned,
			"assigned_agent": agent,
			"updated_at":    time.Now(),
		}).Error
}

func (r *TaskRepository) IncrementRetry(taskID string) error {
	return database.GetDB().Model(&models.ProbeTask{}).
		Where("id = ?", taskID).
		Updates(map[string]interface{}{
			"retry_count": database.GetDB().Raw("retry_count + 1"),
			"updated_at":  time.Now(),
		}).Error
}

func (r *TaskRepository) ListByEnvID(envID string, page, pageSize int) ([]models.ProbeTask, int64, error) {
	var tasks []models.ProbeTask
	var total int64
	db := database.GetDB().Model(&models.ProbeTask{}).Where("env_id = ?", envID)
	db.Count(&total)
	offset := (page - 1) * pageSize
	if err := db.Offset(offset).Limit(pageSize).Order("created_at DESC").Find(&tasks).Error; err != nil {
		return nil, 0, err
	}
	return tasks, total, nil
}

func (r *TaskRepository) ListByStatus(status string, page, pageSize int) ([]models.ProbeTask, int64, error) {
	var tasks []models.ProbeTask
	var total int64
	db := database.GetDB().Model(&models.ProbeTask{})
	if status != "" {
		db = db.Where("status = ?", status)
	}
	db.Count(&total)
	offset := (page - 1) * pageSize
	if err := db.Offset(offset).Limit(pageSize).Order("created_at DESC").Find(&tasks).Error; err != nil {
		return nil, 0, err
	}
	return tasks, total, nil
}

func (r *TaskRepository) GetTaskHistory(envID string, startTime, endTime time.Time) ([]models.ProbeTask, error) {
	var tasks []models.ProbeTask
	db := database.GetDB().Where("env_id = ? AND created_at >= ? AND created_at <= ?", envID, startTime, endTime)
	if err := db.Order("created_at DESC").Find(&tasks).Error; err != nil {
		return nil, err
	}
	return tasks, nil
}

func (r *TaskRepository) Delete(id string) error {
	return database.GetDB().Transaction(func(tx *gorm.DB) error {
		if err := tx.Delete(&models.NetworkResult{}, "task_id = ?", id).Error; err != nil {
			return err
		}
		if err := tx.Delete(&models.DNSRecord{}, "task_id = ?", id).Error; err != nil {
			return err
		}
		if err := tx.Delete(&models.DiagnosisConclusion{}, "task_id = ?", id).Error; err != nil {
			return err
		}
		if err := tx.Delete(&models.ProbeTask{}, "id = ?", id).Error; err != nil {
			return err
		}
		return nil
	})
}
